src/abe_policy/policy.rs

```diff
@@ -244,12 +225,94 @@ impl Policy {
         Ok(())
     }
 
+    /// Adds the given policy axis to the policy.
+    pub fn remove_axis(&mut self, axis_name: String) -> Result<(), Error> {
+        match self.axes.get(&axis_name) {
+            Some(policy_axis) => {
+                if self.axes.len() == 1 {
+                    return Err(Error::UnsupportedOperator(
+                        "Cannot remove the last axis from this policy".to_string(),
+                    ));
+                }
+                for attr_name in &policy_axis.attribute_names {
+                    let attr = Attribute::new(&axis_name, attr_name);
+                    if self.attributes.get(&attr).is_none() {
+                        return Err(Error::AttributeNotFound(format!("{attr:?}")));
+                    }
+                    self.attributes.remove(&attr);
+                }
+                self.axes.remove(&axis_name);
+                Ok(())
+            }
+            None => Err(Error::AxisNotFound(axis_name)),
+        }
+    }
+
+    pub fn add_attribute(
+        &mut self,
+        attr: Attribute,
+        encryption_hint: EncryptionHint,
+    ) -> Result<(), Error> {
+        if self.attributes.get(&attr).is_some() {
+            return Err(Error::ExistingPolicy(format!("{attr:?}")));
+        }
+
+        match self.axes.get(&attr.axis) {
+            Some(policy_axis) => {
+                self.last_attribute_value += 1;
+                let attribute_parameters = PolicyAttributesParameters {
+                    values: [self.last_attribute_value].into(),
+                    encryption_hint,
+                };
+                self.attributes.insert(attr.clone(), attribute_parameters);
+
+                let mut updated_policy_axis = policy_axis.clone();
+                updated_policy_axis.attribute_names.push(attr.name);
+                self.axes.insert(attr.axis, updated_policy_axis);
+
+                Ok(())
+            }
+            None => Err(Error::AxisNotFound(attr.axis)),
+        }
+    }
+
+    pub fn remove_attribute(&mut self, attr: Attribute) -> Result<(), Error> {
+        match self.axes.get(&attr.axis) {
+            Some(policy_axis) => {
+                if self.attributes.get(&attr).is_none() {
+                    return Err(Error::AttributeNotFound(format!("{attr:?}")));
+                }
+
+                let mut updated_policy_axis = policy_axis.clone();
+                let index = updated_policy_axis
+                    .attribute_names
+                    .iter()
+                    .position(|s| s == &attr.name)
+                    .ok_or(Error::AttributeNotFound(attr.name.clone()))?;
+                updated_policy_axis.attribute_names.remove(index);
```

Comment:
In order to avoid this circumvented way of removing attribute names and, it would make sense to:
- use a set to store attribute names for non-hierarchical axes;
- use a sequence (linked list?) to store attribute names for hierarchical axes.

It would also make the structure arguably clearer.